### DISCUSSION TITLE
Feature: support `linux/s390x`

### DISCUSSION THREAD

**Discussion by skuethe** [fluxcd/flux2#2475] [Proposals] [answered]
Hello everybody,

I am currently in the process of adding PRs to the various controllers to support `linux/s390x` platform builds.
Before putting in all the work, I would like to sync with the maintainers if this is an accepted change / feature enhancement.

As multi arch images are setup already (for arm, arm64 and amd64), changes should be mostly small (without me having them tested yet).

Nevertheless, maybe there is some culprit I don't yet know about which would make this work redundant (like specific packages or builds not being available for s390x). If you have some feedback there, let me know :)

I have already started with flux2 and helm-controller PRs:
- https://github.com/fluxcd/flux2/pull/2471
- https://github.com/fluxcd/helm-controller/pull/426

Thanks and kind regards

**Comment by kingdonb** [comment]
I don't know nearly enough to say what type of issues you may encounter (or already have discovered).

But the places I expect you'll be more likely to run into trouble are the source-controller and image-automation-controller, as I understand a lot of work has gone into making the static builds work successfully to enable the fuzzing support, and to make them multi-arch capable while also doing static builds. (I, for one, definitely look forward to seeing Flux running on mainframes. 🎉)

My biggest concern about accepting an s390x port is how we will test it in the future. The Flux project has limited resources but is able to test arm64 builds thanks to donated hardware and funding for hosting it that comes, as I understand it, through the CNCF.

We have redone our build architecture several times already and each time, without dedicated resources to test the change, it is harder and harder to verify that we have not broken anything on any architecture. Do you have any proposal or suggestion for how we can approach that issue? (Edit: I see you have thought about this, and it's mentioned on your PR.) Thanks again for proposing this.

**Comment by stefanprodan** [comment, marked as answer]
Adding a new arch to Flux is no easy feat because source-controller depends on C libraries such as libgit2, libssh2, openssl, etc. Before we can consider `linux/s390x`, the musl static build must be implemented in https://github.com/fluxcd/golang-with-libgit2

**Reply by skuethe** [reply]
Okay, just saw this. Thanks, I will have a look if this is anything realistic to do.

**Reply by pjbgf** [reply]
@skuethe on a quick test trying to get https://github.com/fluxcd/golang-with-libgit2 to build for `linux/s390x` I noticed that it may not be as straight-forward as the existing architectures:
- `lld` does not seem to be available on the platform specific alpine index.
- Building openssl for the `linux32-s390x` target does not generate a clean build - using buildx with `qemu-s390x` installed.

It may require a few changes on the way we build the dependency chain for this to work.
For us to add this as officially support, we will need to make sure we can test and ensure all the features are working to the same standard as we do for the other architectures, as kingdonb mentioned above. 

If you are still keen to give it a go, first steps would be to add a new trimmed layer for that architecture in the [Dockerfile](https://github.com/fluxcd/golang-with-libgit2/blob/main/Dockerfile#L81):
```
FROM scratch as libs-s390x
COPY --from=trimmed /trimmed/ /
```

Then try to build the project with:
`PLATFORMS=linux/s390x make build`

From that point you can iterate through the errors until you have a working version. Once that is building successfully, you can amend the [Dockerfile.test](https://github.com/fluxcd/golang-with-libgit2/blob/main/Dockerfile.test) accordingly and run `PLATFORMS=linux/s390x make test` to ensure that the dependency chain is actually working properly.

**Reply by skuethe** [reply]
Thank you for having a look at this already! I need to find some time to evaluate this :)
